File: server/src/services/paymentService.js
import crypto from "crypto";
import { Order } from "../models/Order.js";
import { Payment } from "../models/Payment.js";
import { ApiError } from "../utils/ApiError.js";

export const createPaymentIntent = async (userId, orderId) => {
  const order = await Order.findOne({ _id: orderId, userId }).lean();
  if (!order) throw new ApiError("Order not found", 404);

  const payment = await Payment.create({
    orderId,
    userId,
    amount: order.total,
    transactionId: crypto.randomUUID(),
    status: "pending",
  });

  return payment.toObject();
};

export const verifyPayment = async (transactionId, status) => {
  const payment = await Payment.findOne({ transactionId });
  if (!payment) throw new ApiError("Payment not found", 404);

  payment.status = status;
  await payment.save();

  const isPaid = status === "succeeded";
  await Order.updateOne(
    { _id: payment.orderId },
    { paymentStatus: isPaid ? "paid" : "failed", isPaid },
  );

  return payment.toObject();
};

File: server-spring/src/main/java/com/ecommerce/modules/payment/service/PaymentService.java
package com.ecommerce.modules.payment.service;

import com.ecommerce.modules.order.model.Order;
import com.ecommerce.modules.order.model.PaymentStatus;
import com.ecommerce.modules.order.repository.OrderRepository;
import com.ecommerce.modules.payment.dto.CreatePaymentRequest;
import com.ecommerce.modules.payment.dto.VerifyPaymentRequest;
import com.ecommerce.modules.payment.model.Payment;
import com.ecommerce.modules.payment.repository.PaymentRepository;
import com.ecommerce.shared.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class PaymentService {

    private final PaymentRepository paymentRepository;
    private final OrderRepository orderRepository;

    public Payment createPaymentIntent(Long userId, CreatePaymentRequest request) {
        Order order = orderRepository.findById(request.getOrderId())
                .filter(o -> o.getUserId().equals(userId))
                .orElseThrow(() -> new ResourceNotFoundException("Order not found"));

        Payment payment = Payment.builder()
                .orderId(order.getId())
                .userId(userId)
                .amount(order.getTotal())
                .transactionId(UUID.randomUUID().toString())
                .status("pending")
                .build();

        return paymentRepository.save(payment);
    }

    @Transactional
    public Payment verifyPayment(VerifyPaymentRequest request) {
        Payment payment = paymentRepository.findByTransactionId(request.getTransactionId())
                .orElseThrow(() -> new ResourceNotFoundException("Payment not found"));

        payment.setStatus(request.getStatus());
        paymentRepository.save(payment);

        boolean isPaid = "succeeded".equals(request.getStatus());
        Order order = orderRepository.findById(payment.getOrderId())
                .orElseThrow(() -> new ResourceNotFoundException("Order not found"));
        order.setPaymentStatus(isPaid ? PaymentStatus.PAID : PaymentStatus.FAILED);
        order.setIsPaid(isPaid);
        orderRepository.save(order);

        return payment;
    }
}
